pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -92,6 +101,45 @@ type GraphBuilder struct {
 	absentOwnerCache *UIDCache
 }
 
+type stopMux struct {
+	// if either globalStop or localStop is closed, outputStop will be closed.
+	globalStop <-chan struct{}
+	localStop  chan struct{}
+	outputStop chan struct{}
+}
+
+func newStopMux(stop <-chan struct{}) *stopMux {
+	sm := &stopMux{
+		globalStop: stop,
+		localStop:  make(chan struct{}),
+		outputStop: make(chan struct{}),
+	}
+	go sm.stopper()
+	return sm
+}
+
+func (sm *stopMux) stopper() {
+	defer close(sm.outputStop)
+	select {
+	case <-sm.globalStop:
+	case <-sm.localStop:
+	}
+}
+
+type monitorSingle struct {
```

Comment:
```
// monitorSingle wraps a cache.Controller and adds another stop channel (in addition to that of Run(stopCh)). It will stop
// if either of them is closed. 
```